src/ResourceManager/Common/Commands.Common.Strategies/ResourceStrategy.cs

```diff
@@ -0,0 +1,111 @@
+        { get; }
+
+        public Func<TModel, string> GetLocation { get; }
+
+        public Action<TModel, string> SetLocation { get; }
+
+        public Func<TModel, int> CreateTime { get; }
+
+        public bool CompulsoryLocation { get; }
+
+        public ResourceStrategy(
+            string type,
+            Func<string, IEnumerable<string>> getId,
+            Func<IClient, GetAsyncParams, Task<TModel>> getAsync,
+            Func<IClient, CreateOrUpdateAsyncParams<TModel>, Task<TModel>> createOrUpdateAsync,
+            Func<TModel, string> getLocation,
+            Action<TModel, string> setLocation,
+            Func<TModel, int> createTime,
+            bool compulsoryLocation)
+        {
+            Type = type;
+            GetId = getId;
+            GetAsync = getAsync;
+            CreateOrUpdateAsync = createOrUpdateAsync;
+            GetLocation = getLocation;
+            SetLocation = setLocation;
+            CreateTime = createTime;
+            CompulsoryLocation = compulsoryLocation;
+        }
+    }
+
+    public static class ResourceStrategy
+    {
+        public static ResourceStrategy<TModel> Create<TModel, TClient, TOperation>(
+            string type,
+            Func<string, IEnumerable<string>> getId,
+            Func<TClient, TOperation> getOperations,
+            Func<TOperation, GetAsyncParams, Task<TModel>> getAsync,
+            Func<TOperation, CreateOrUpdateAsyncParams<TModel>, Task<TModel>> createOrUpdateAsync,
+            Func<TModel, string> getLocation,
+            Action<TModel, string> setLocation,
+            Func<TModel, int> createTime,
+            bool compulsoryLocation)
+            where TClient : ServiceClient<TClient>
+        {
+            Func<IClient, TOperation> toOperations = client => getOperations(client.GetClient<TClient>());
+            return new ResourceStrategy<TModel>(
+                type,
+                getId,
+                (client, p) => getAsync(toOperations(client), p),
+                (client, p) => createOrUpdateAsync(toOperations(client), p),
+                getLocation,
+                setLocation,
+                createTime,
+                compulsoryLocation);
+        }
+
+        public static ResourceStrategy<TModel> Create<TModel, TClient, TOperation>(
+            string type,
+            IEnumerable<string> headers,
+            Func<TClient, TOperation> getOperations,
+            Func<TOperation, GetAsyncParams, Task<TModel>> getAsync,
+            Func<TOperation, CreateOrUpdateAsyncParams<TModel>, Task<TModel>> createOrUpdateAsync,
+            Func<TModel, string> getLocation,
+            Action<TModel, string> setLocation,
+            Func<TModel, int> createTime,
+            bool compulsoryLocation)
+            where TClient : ServiceClient<TClient>
+            => Create(
+                type,
+                name => new[] { "providers" }.Concat(headers).Concat(new[] { name }),
```

Comment:
>headers [](start = 53, length = 7)

this should be changed to "providers". I did revert my changes due to conflicts :)